primitives/utility/src/lib.rs

```diff
@@ -51,3 +61,157 @@ impl<T: UpwardMessageSender, W: WrapVersion> SendXcm for ParentAsUmp<T, W> {
 		}
 	}
 }
+
+pub struct TakeFirstAssetTrader<
+	AccountId,
+	FeeCharger: ChargeWeightInFungibles<AccountId, Assets>,
+	Matcher: MatchesFungibles<Assets::AssetId, Assets::Balance>,
+	Assets: fungibles::Mutate<AccountId> + fungibles::Transfer<AccountId> + fungibles::Balanced<AccountId>,
+	HandleRefund: TakeRevenue,
+>(
+	Weight,
+	Assets::Balance,
+	Option<(MultiLocation, Assets::AssetId)>,
+	PhantomData<(AccountId, FeeCharger, Matcher, Assets, HandleRefund)>,
+);
+impl<
+		AccountId,
+		FeeCharger: ChargeWeightInFungibles<AccountId, Assets>,
+		Matcher: MatchesFungibles<Assets::AssetId, Assets::Balance>,
+		Assets: fungibles::Mutate<AccountId>
+			+ fungibles::Transfer<AccountId>
+			+ fungibles::Balanced<AccountId>,
+		HandleRefund: TakeRevenue,
+	> WeightTrader for TakeFirstAssetTrader<AccountId, FeeCharger, Matcher, Assets, HandleRefund>
+{
+	fn new() -> Self {
+		Self(0, Zero::zero(), None, PhantomData)
+	}
+	// We take first asset
+	// TODO: do we want to be iterating over all potential assets in payments?
+	// Check whether we can convert fee to asset_fee (is_sufficient, min_deposit)
+	// If everything goes well, we charge.
+	fn buy_weight(
+		&mut self,
+		weight: Weight,
+		payment: xcm_executor::Assets,
+	) -> Result<xcm_executor::Assets, XcmError> {
+		log::trace!(target: "xcm::weight", "TakeFirstAssetTrader::buy_weight weight: {:?}, payment: {:?}", weight, payment);
+		// We take the very first multiasset from payment
+		// TODO: revisit this clone
+		let multiassets: MultiAssets = payment.clone().into();
+		let first = multiassets.get(0).ok_or(XcmError::TooExpensive)?;
+		// We get the local asset id in which we can pay for fees
+		let (local_asset_id, _) =
+			Matcher::matches_fungibles(&first).map_err(|_| XcmError::TooExpensive)?;
```

Comment:
I think it is more informative and accurate `AssetNotFound`
```suggestion
			Matcher::matches_fungibles(&first).map_err(|_| XcmError::AssetNotFound)?;
```